### PR TITLE
✨ [source-google-sheets] add row_batch_size as an input parameter with higher increase

### PR DESCRIPTION
**Pull request by tautvydas-v**
## What
Accidentally closed [#35320](https://github.com/airbytehq/airbyte/pull/35320) PR therefore raising this new one.
Resolves my raised issue: https://github.com/airbytehq/airbyte/issues/35274
Previously, source_google_sheets source connector had hardcoded value of 200 for row_batch_size. This means that one request sent to Google Sheets API processes only 200 rows, even though there's pretty much no limit, as long as it's processed in under 3 minutes. Also, limitation is on requests sent, but not the rows processed. Sometimes, a problem rises when doing a sync with a google sheet with over 100k records, where exponential backoff fails and the sync silently fails. In order to overcome this, making "batch_size" input parameter, with a default value of 200, so there wouldn't be any kind of change when updating the connector.

## How
Added "batch_size" to spec with a default value of 200 which is passed over **source.py** file, under **_read** method. Also, changed row_batch_size increase to 100 if the exception code is _TOO_MANY_REQUESTS_.

### CODE REVIEW

**Comment by darynaishchenko** [comment]
@tautvydas-v, Regarding this comment in previous pr https://github.com/airbytehq/airbyte/pull/35320#issuecomment-1948368756
What is suitable input row batch size value for you? I'm ok with idea of percentage, but I'm afraid that not all users will understand the logic of this input param and they should take some time to understand what value they need.  Did you test this changes with suitable row batch size value and backoff logic on your data?

**Comment by tautvydas-v** [comment]
Hey @darynaishchenko, suitable batch size really depends case on case I suppose. In some cases, we could really use 10k records per batch and it would speed the extraction by a lot. Keep in mind that we're leaving default batch size as 200 and it's under _Optional_ parameters, so even with a version update, if someone doesn't really need or care about batch size - it won't have any impact, as it's already currently 200. As a use case, we have a google sheet with about ~137000 records and here's a screenshot of how it goes for us so far:
Screenshot 2024-02-21 at 12 09 56 PM
As you can see, in some cases we extract 30k-80k records even though we have 137k records. So in a way, it silently fails. Also, it takes up to 10 minutes for full extraction.

Meanwhile, we've tested out with different batch size inputs:
Here's with 10k rows per batch:
Screenshot 2024-02-21 at 12 40 48 PM
Screenshot 2024-02-21 at 12 42 25 PM


Here's with 100k rows per batch:
Screenshot 2024-02-21 at 1 02 25 PM
Screenshot 2024-02-21 at 12 52 50 PM

Do you see any potential problems with this setup though? Having 100k rows as a batch size really seems to be an edge case, but it's possible to use it too.

**Comment by tautvydas-v** [comment]
@darynaishchenko @marcosmarxm Hi, what should be the next steps here?

**Comment by marcosmarxm** [comment]
@darynaishchenko can you review the contributio again please?

**Comment by tautvydas-v** [comment]
@darynaishchenko @marcosmarxm Hey, any additional steps should be taken here? Just conscious that this doesn't get lost as we would really like to implement this.

File: airbyte-integrations/connectors/source-google-sheets/source_google_sheets/spec.yaml [hunk 1]
```diff
@@ -8,6 +8,12 @@ connectionSpecification:
     - credentials
   additionalProperties: true
   properties:
+    batch_size:
+      type: integer
+      title: Row Batch Size
+      description: >-
+        An integer representing row batch size with each sent request to Google Sheets API. Defaults to 200.
```
Comment (darynaishchenko) [inline airbyte-integrations/connectors/source-google-sheets/source_google_sheets/spec.yaml:15]: @tautvydas-v lgtm, could you please add more detailed description here and in .md file for Row Batch Size. Info about what impact this value have on rate limits, sycn etc, and how to define most suitable value for users.

File: airbyte-integrations/connectors/source-google-sheets/source_google_sheets/source.py [hunk 2]
```diff
@@ -149,6 +149,7 @@ def _read(
         catalog: ConfiguredAirbyteCatalog,
     ) -> Generator[AirbyteMessage, None, None]:
         client = GoogleSheetsClient(self.get_credentials(config))
+        client.Backoff.row_batch_size = config["batch_size"]
```
Comment (darynaishchenko) [inline airbyte-integrations/connectors/source-google-sheets/source_google_sheets/source.py:152]:
```suggestion
        client.Backoff.row_batch_size = config.get("batch_size", 200)
```
for old configs without batch_size.

File: airbyte-integrations/connectors/source-google-sheets/source_google_sheets/spec.yaml (hunk 1 above)
Comment (tautvydas-v) [reply airbyte-integrations/connectors/source-google-sheets/source_google_sheets/spec.yaml]: Added in more documentation

File: airbyte-integrations/connectors/source-google-sheets/source_google_sheets/source.py (hunk 2 above)
Comment (tautvydas-v) [reply airbyte-integrations/connectors/source-google-sheets/source_google_sheets/source.py]: Fixed

**Comment by marcosmarxm** [comment]
@darynaishchenko let me know if this is ready to ship 🚀

**Comment by tautvydas-v** [comment]
@darynaishchenko Hi, what's the status with this PR? Can this be reviewed again?

**Comment by marcosmarxm** [comment]
@tautvydas-v there are some internal discussion about the implementation, sorry the delay. Hope to get an update next week! Thanks for the your patience.

**Comment by tautvydas-v** [comment]
@marcosmarxm @darynaishchenko Hi, any news regarding this one? I've noticed that I have to update the connector version for the fourth time, but really not sure what's the status here and whether it's even worth it to do so in the future. Would be great to get some feedback, thanks.

**Comment by darynaishchenko** [comment]
@tautvydas-v Hi, I'm ok with your solution. 
Just one last question, what do you think about calculating of proper batch size before actual sync by fetching a count of records in spread sheet? So for example before read we fetch that request spread sheet has 100k records and set row_batch size to 10 000 as you used. And we can use enum for row_batch size value for example sheets with < 10000 records: batch size is 200, > 10000 and  < 100 000:   batch size is 1000 etc.

**Comment by tautvydas-v** [comment]
Hey @darynaishchenko,

In my opinion, even though setting `batch_size` based on Google sheet size seems like a more dynamic solution, I'm a bit conscious about debugging as an end user, more specifically - hitting API timeout limits (timeouts after 180 seconds I believe). Not sure how often that can happen but if a user has a slow connection, has a huge Google sheet and tries to sync the sheet with dynamically set batch size (which could be for example 5-10k records if a sheet is really big) and gets a timeout error - it can be confusing why that happened.

Having default batch_size as 200 seems to cover a lot of cases, is more clear when having lots of connections and makes checking logs a bit more clearer due to the fact that you can see the same value for each sync. If a user wants to just use a connection and will see different values for each sync (where the Google sheets size is different), I think this could create confusion. 
Also, one source accepts one google sheet (although number of sheets inside the Google sheet could vary), so it should be clear enough to notice how many rows are in the sheet and decide for a fitting batch_size.

**Review by darynaishchenko — APPROVED** [review]
lgtm
@tautvydas-v thanks for your work on it.
@marcosmarxm it's ready to ship

**Comment by marcosmarxm** [comment]
Thanks @darynaishchenko. @tautvydas-v I'm going to proceed with the shipping process 🚢

**Pull request merged**